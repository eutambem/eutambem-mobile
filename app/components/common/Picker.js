import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { View, TextInput, StyleSheet, Picker as PickerAndroid, PickerIOS, Platform } from 'react-native';
import Modal from 'react-native-modal';

import appStyles from '../../styles';
import Button from '../common/Button';

const styles = StyleSheet.create({
  modalContainer: {
    flex: 1,
    justifyContent: 'flex-end',
    margin: 0,
  },
  innerContainer: {
    backgroundColor: 'white',
    opacity: 1,
    borderRadius: 4,
    borderColor: 'rgba(0, 0, 0, 0.1)',
  },
});

export default class Picker extends Component {
  constructor(props) {
    super(props);
    this.state = {
      modalVisible: false,
      itemSelected: {},
      currentItem: {},
    };
    this.onValueChange = this.onValueChange.bind(this);
  }

  onValueChange(itemValue) {
    const itemLabel = this.props.items.find(item => item.value === itemValue).label;
    this.setState({ currentItem: { value: itemValue, label: itemLabel } });
  }

  confirm() {
    this.setState({ itemSelected: this.state.currentItem });
    this.closeModal();
  }

  openModal() {
    this.input.blur();
    this.setState({ modalVisible: true });
  }

  closeModal() {
    this.setState({ modalVisible: false });
  }

  inputRef(component) { this.input = component; }

  render() {
    return Platform.select({
      ios: (
        <View>
          <TextInput
            style={appStyles.input}
            placeholder={this.props.placeholder}
            value={this.state.itemSelected.label}
            onFocus={() => this.openModal()}
            ref={this.inputRef.bind(this)}
          />
          <Modal
            transparent
            isVisible={this.state.modalVisible}
            style={styles.modalContainer}
            onRequestClose={() => this.closeModal()}
          >
            <View style={styles.innerContainer}>
              <View style={appStyles.modalHeader}>
                <Button
                  type="cancel"
                  style={[appStyles.button]}
                  onPress={() => this.closeModal()}
                  title="Cancelar"
                />
                <Button
                  style={appStyles.button}
                  onPress={() => this.confirm()}
                  title="Ok"
                />
              </View>
              <View>
                <PickerIOS
                  style={{ width: '100%' }}
                  selectedValue={this.state.currentItem.value}
                  onValueChange={this.onValueChange}
                >
                  { this.props.items.map(item => (
                    <PickerIOS.Item
                      key={item.value}
                      label={item.label}
                      value={item.value}
                    />
                  ))}
                </PickerIOS>
              </View>
            </View>
          </Modal>
        </View>
      ),
      android: (
        <PickerAndroid
          style={{ width: '100%' }}
          placeholder={this.props.placeholder}
          selectedValue={this.state.currentItem.value}
          onValueChange={this.onValueChange}
        >
          { this.props.items.map(item => (
            <PickerAndroid.Item
              key={item.value}
              label={item.label}
              value={item.value}
            />
          ))}
        </PickerAndroid>
      ),
    });
  }
}

Picker.propTypes = {
  items: PropTypes.arrayOf(PropTypes.shape({
    label: PropTypes.string,
    value: PropTypes.string,
  })).isRequired,
  placeholder: PropTypes.string.isRequired,
};

