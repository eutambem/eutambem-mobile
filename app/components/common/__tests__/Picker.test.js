import 'react-native';
import React from 'react';
import { shallow } from 'enzyme';

import Picker from '../Picker';

const props = {
  placeholder: 'Test Picker',
  items: [{
    value: '1',
    label: 'Label 1',
  }, {
    value: '2',
    label: 'Label 2',
  }],
};

describe('Picker', () => {
  it('renders correctly', () => {
    const wrapper = shallow(<Picker {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
});
