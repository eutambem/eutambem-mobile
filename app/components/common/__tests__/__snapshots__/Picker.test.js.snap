// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Picker renders correctly 1`] = `
<ForwardRef(View)>
  <TextInput
    multiline={false}
    onFocus={[Function]}
    placeholder="Test Picker"
    required={false}
  />
  <ReactNativeModal
    animationIn="slideInUp"
    animationInTiming={300}
    animationOut="slideOutDown"
    animationOutTiming={300}
    avoidKeyboard={false}
    backdropColor="black"
    backdropOpacity={0.7}
    backdropTransitionInTiming={300}
    backdropTransitionOutTiming={300}
    hideModalContentWhileAnimating={false}
    isVisible={false}
    onBackButtonPress={[Function]}
    onBackdropPress={[Function]}
    onModalHide={[Function]}
    onModalShow={[Function]}
    onRequestClose={[Function]}
    style={
      Object {
        "flex": 1,
        "justifyContent": "flex-end",
        "margin": 0,
      }
    }
    swipeThreshold={100}
    transparent={true}
    useNativeDriver={false}
  >
    <ForwardRef(View)
      style={
        Object {
          "backgroundColor": "white",
          "borderColor": "rgba(0, 0, 0, 0.1)",
          "borderRadius": 4,
          "opacity": 1,
        }
      }
    >
      <ForwardRef(View)
        style={
          Object {
            "alignItems": "center",
            "borderColor": "#eeeeee",
            "borderWidth": 1,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "padding": 3,
          }
        }
      >
        <Button
          onPress={[Function]}
          title="Cancelar"
          type="cancel"
        />
        <Button
          onPress={[Function]}
          title="Ok"
          type="confirm"
        />
      </ForwardRef(View)>
      <ForwardRef(View)>
        <PickerIOS
          onValueChange={[Function]}
          selectedValue="1"
          style={
            Object {
              "width": "100%",
            }
          }
        >
          <PickerIOSItem
            key="1"
            label="Label 1"
            value="1"
          />
          <PickerIOSItem
            key="2"
            label="Label 2"
            value="2"
          />
        </PickerIOS>
      </ForwardRef(View)>
    </ForwardRef(View)>
  </ReactNativeModal>
</ForwardRef(View)>
`;
